# Digital Music Concert Archive (DMCA) Data Anlysis

From the period of ~2005 to now, the concert archives have been recorded in a pseudo-database within FileMaker Pro. Unfortunately, since this is not a standard database, we can't use robust tools within python that handle SQL (at least not without some conversion).  

For now, though, we can get the tables into data frames for `pandas` analysis. Since one or two of the tables are incomplete, we can start with the "contributors" table. Instead of what you would expect, where one performer gets one unique (primary) key, a new key was produced for _each_ performer, for _each_ piece.  

We could handle this in a couple different ways, but since the main table that we are interested in, the "pieces" table, has keys that correspond to the "performers" table via the key "Piece_ID", we could try to normalize on that, producing a table that has all the Piece IDs, with each row then having a list of performers.   

In [2]:
import pandas as pd

Let's turn the csv into a dataframe:

In [3]:
%%time
contrib_data = 'ca_contrib-cleanup.csv'
df = pd.read_csv(contrib_data, index_col='Contrib_ID')

CPU times: user 16 ms, sys: 20 ms, total: 36 ms
Wall time: 51.1 ms


Let's look at a slice of the table:

In [4]:
df[:9]

Piece_ID     Instrument              Name
Contrib_ID                                             
1           2009102501          Cello     Erenyi, Erica
2           2009102501          Piano       Lee, Yvonne
3           2009102501         Violin    Négyesy, János
4           2009102501  Violin, Viola  Nykter, Päivikki
5           2009102501  Counter Tenor      Toledo, John
6           2009102502          Cello     Erenyi, Erica
7           2009102502          Piano       Lee, Yvonne
8           2009102502         Violin    Négyesy, János
9           2009102502  Violin, Viola  Nykter, Päivikki

## Fundamental Database Issues

Note how for the performer "Negyesy, Janos", he has two unique Contributor IDs. It turns out he has many more in this table. We need to get around that, so as mentioned above, let's just use Piece ID to get a list of performers for each piece. We can do this via a self-join or pivot.  

However, if we try to pivot on Piece ID, `pandas` will complain the index contains duplicates. To get around this, we need to make an index on Piece ID that groups them into unique groups:

In [5]:
df['idx'] = df.groupby('Piece_ID').cumcount()

Now we have that unique index of our piece IDs, we can merge on those to just get the names for each piece:

In [6]:
df_norm = df.pivot(index='Piece_ID', columns='idx')[['Name']]
df_norm[:15]

Name                                \
idx                                 0                             1    
Piece_ID                                                               
2007002                 Ohnishi, Takae                          None   
2007003                 Ohnishi, Takae                          None   
2007004                 Ohnishi, Takae                          None   
2007005                 Ohnishi, Takae                          None   
2007006                 Ohnishi, Takae                          None   
2007007                 Ohnishi, Takae                          None   
2007008   La Jolla Symphony and Chorus                          None   
2007009                  Sutter, Wendy  La Jolla Symphony and Chorus   
2007010   La Jolla Symphony and Chorus                          None   
2007011              Zelickman, Robert                          None   
2007012              Zelickman, Robert                          None   
2007013              Zelickman, Robert                          None   
2007014              Zelickman, Robert               Oliveira, Fabio   
2007015                Gattegno, Eliot                 Wubbels, Eric   
2007016                Gattegno, Eliot                 Wubbels, Eric   

                                                                            \
idx                  2                  3             4     5     6     7    
Piece_ID                                                                     
2007002            None               None          None  None  None  None   
2007003            None               None          None  None  None  None   
2007004            None               None          None  None  None  None   
2007005            None               None          None  None  None  None   
2007006            None               None          None  None  None  None   
2007007            None               None          None  None  None  None   
2007008            None               None          None  None  None  None   
2007009            None               None          None  None  None  None   
2007010            None               None          None  None  None  None   
2007011            None               None          None  None  None  None   
2007012            None               None          None  None  None  None   
2007013            None               None          None  None  None  None   
2007014   Dresser, Mark  McAllister, Colin  Menezes, Flo  None  None  None   
2007015            None               None          None  None  None  None   
2007016            None               None          None  None  None  None   

                      ...                                                   \
idx         8     9   ...     37    38    39    40    41    42    43    44   
Piece_ID              ...                                                    
2007002   None  None  ...   None  None  None  None  None  None  None  None   
2007003   None  None  ...   None  None  None  None  None  None  None  None   
2007004   None  None  ...   None  None  None  None  None  None  None  None   
2007005   None  None  ...   None  None  None  None  None  None  None  None   
2007006   None  None  ...   None  None  None  None  None  None  None  None   
2007007   None  None  ...   None  None  None  None  None  None  None  None   
2007008   None  None  ...   None  None  None  None  None  None  None  None   
2007009   None  None  ...   None  None  None  None  None  None  None  None   
2007010   None  None  ...   None  None  None  None  None  None  None  None   
2007011   None  None  ...   None  None  None  None  None  None  None  None   
2007012   None  None  ...   None  None  None  None  None  None  None  None   
2007013   None  None  ...   None  None  None  None  None  None  None  None   
2007014   None  None  ...   None  None  None  None  None  None  None  None   
2007015   None  None  ...   None  None  None  None  None  None  None  None   
200701

So now that we concatenated vertically, we see there's an interesting issue... some of these pieces have a whole lot of performers. In fact, from the above output we now know there's a piece with 46 performers!

## Aligning with our template

Our data is in a decent state where we could normally merge this dataframe with the other table, then take that into OpenRefine for some cleaning and be done. However, there is a very specific metadata template that this data will be going into. Part of that specificity is that multi-valued cells are delimited with a "space pipe space" (` | `). So, we should concatenate all the contributors in their columns separated by spaces and pipes.  

In [7]:
df_norm['merged_names'] = df_norm.apply(lambda x: ' | '.join(x.dropna().values.tolist()), axis=1)
df_norm[:15]

Name                                \
idx                                  0                             1   
Piece_ID                                                               
2007002                 Ohnishi, Takae                          None   
2007003                 Ohnishi, Takae                          None   
2007004                 Ohnishi, Takae                          None   
2007005                 Ohnishi, Takae                          None   
2007006                 Ohnishi, Takae                          None   
2007007                 Ohnishi, Takae                          None   
2007008   La Jolla Symphony and Chorus                          None   
2007009                  Sutter, Wendy  La Jolla Symphony and Chorus   
2007010   La Jolla Symphony and Chorus                          None   
2007011              Zelickman, Robert                          None   
2007012              Zelickman, Robert                          None   
2007013              Zelickman, Robert                          None   
2007014              Zelickman, Robert               Oliveira, Fabio   
2007015                Gattegno, Eliot                 Wubbels, Eric   
2007016                Gattegno, Eliot                 Wubbels, Eric   

                                                                            \
idx                   2                  3             4     5     6     7   
Piece_ID                                                                     
2007002            None               None          None  None  None  None   
2007003            None               None          None  None  None  None   
2007004            None               None          None  None  None  None   
2007005            None               None          None  None  None  None   
2007006            None               None          None  None  None  None   
2007007            None               None          None  None  None  None   
2007008            None               None          None  None  None  None   
2007009            None               None          None  None  None  None   
2007010            None               None          None  None  None  None   
2007011            None               None          None  None  None  None   
2007012            None               None          None  None  None  None   
2007013            None               None          None  None  None  None   
2007014   Dresser, Mark  McAllister, Colin  Menezes, Flo  None  None  None   
2007015            None               None          None  None  None  None   
2007016            None               None          None  None  None  None   

                                            ...                                \
idx          8     9                        ...                            38   
Piece_ID                                    ...                                 
2007002   None  None                        ...                          None   
2007003   None  None                        ...                          None   
2007004   None  None                        ...                          None   
2007005   None  None                        ...                          None   
2007006   None  None                        ...                          None   
2007007   None  None                        ...                          None   
2007008   None  None                        ...                          None   
2007009   None  None                        ...                          None   
2007010   None  None                        ...                          None   
2007011   None  None                        ...                          None   
2007012   None  None                        ...                          None   
2007013   None  None                        ...                          None   
2007014   None  None                        ...                          None   
2007015   None  None                

Now we can slice out the unnecessary columns

In [8]:
df1 = df_norm[['merged_names']]
df1[:29]

merged_names
idx                                                        
Piece_ID                                                   
2007002                                      Ohnishi, Takae
2007003                                      Ohnishi, Takae
2007004                                      Ohnishi, Takae
2007005                                      Ohnishi, Takae
2007006                                      Ohnishi, Takae
2007007                                      Ohnishi, Takae
2007008                        La Jolla Symphony and Chorus
2007009        Sutter, Wendy | La Jolla Symphony and Chorus
2007010                        La Jolla Symphony and Chorus
2007011                                   Zelickman, Robert
2007012                                   Zelickman, Robert
2007013                                   Zelickman, Robert
2007014   Zelickman, Robert | Oliveira, Fabio | Dresser,...
2007015                     Gattegno, Eliot | Wubbels, Eric
2007016                     Gattegno, Eliot | Wubbels, Eric
2007017                     Gattegno, Eliot | Wubbels, Eric
2007018                     Gattegno, Eliot | Wubbels, Eric
2007019                     Gattegno, Eliot | Wubbels, Eric
2007020                     Gattegno, Eliot | Wubbels, Eric
2007021                     Gattegno, Eliot | Wubbels, Eric
2007022                         Karis, Aleck | Beck, Steven
2007023                         Karis, Aleck | Beck, Steven
2007024                         Karis, Aleck | Beck, Steven
2007025            DeMaison, Nicholas | Gallagher, Kathleen
2007026                                       Manabe, Reiko
2007027                                          Power, Ian
2007028                                       Tabor, Amanda
2007029                       Manabe, Reiko | Medine, David
2007030                                   McCallum, Clinton

Let's see how many rows we have in total

In [10]:
df1.shape

(5794, 1)

5794 is a bit off from what we're expecting. Let's get a bit more info on the dataframe

In [11]:
df1.info

<bound method DataFrame.info of                                                  merged_names
idx                                                          
Piece_ID                                                     
2007002                                        Ohnishi, Takae
2007003                                        Ohnishi, Takae
2007004                                        Ohnishi, Takae
2007005                                        Ohnishi, Takae
2007006                                        Ohnishi, Takae
2007007                                        Ohnishi, Takae
2007008                          La Jolla Symphony and Chorus
2007009          Sutter, Wendy | La Jolla Symphony and Chorus
2007010                          La Jolla Symphony and Chorus
2007011                                     Zelickman, Robert
2007012                                     Zelickman, Robert
2007013                                     Zelickman, Robert
2007014     Zelickman, Robert | Olivei